File: otelcollector/configmapparser/tomlparser-ccp-default-scrape-settings.rb
#!/usr/local/bin/ruby
# frozen_string_literal: true

require "tomlrb"
require_relative "ConfigParseErrorLogger"

LOGGING_PREFIX = "default-scrape-settings"

@configMapMountPath = "/etc/config/settings/default-scrape-settings-enabled"
@configVersion = ""
@configSchemaVersion = ""

@controlplane_kube_controller_manager_enabled = false
@controlplane_kube_scheduler_enabled = false
@controlplane_apiserver_enabled = true
@controlplane_cluster_autoscaler_enabled = false
@controlplane_etcd_enabled = true
@controleplane_prometheuscollectorhealth_enabled = false
@noDefaultsEnabled = false

# Use parser to parse the configmap toml file to a ruby structure
def parseConfigMap
  begin
    # Check to see if config map is created
    if (File.file?(@configMapMountPath))
      parsedConfig = Tomlrb.load_file(@configMapMountPath, symbolize_keys: true)
      return parsedConfig
    else
      ConfigParseErrorLogger.logWarning(LOGGING_PREFIX, "configmapprometheus-collector-configmap for scrape targets not mounted, using defaults")
      return nil
    end
  rescue => errorStr
    ConfigParseErrorLogger.logError(LOGGING_PREFIX, "Exception while parsing config map for default scrape settings: #{errorStr}, using defaults, please check config map for errors")
    return nil
  end
end

# Use the ruby structure created after config parsing to set the right values to be used for otel collector settings
def populateSettingValuesFromConfigMap(parsedConfig)
  begin
    if !parsedConfig[:"controlplane-kube-controller-manager"].nil?
      @controlplane_kube_controller_manager_enabled = parsedConfig[:"controlplane-kube-controller-manager"]
      puts "config::Using configmap scrape settings for controlplane-kube-controller-manager: #{@controlplane_kube_controller_manager_enabled}"
    end
    if !parsedConfig[:"controlplane-kube-scheduler"].nil?
      @controlplane_kube_scheduler_enabled = parsedConfig[:"controlplane-kube-scheduler"]
      puts "config::Using configmap scrape settings for controlplane-kube-scheduler: #{@controlplane_kube_scheduler_enabled}"
    end
    if !parsedConfig[:"controlplane-apiserver"].nil?
      @controlplane_apiserver_enabled = parsedConfig[:"controlplane-apiserver"]
      puts "config::Using configmap scrape settings for controlplane-apiserver: #{@controlplane_apiserver_enabled}"
    end
    if !parsedConfig[:"controlplane-cluster-autoscaler"].nil?
      @controlplane_cluster_autoscaler_enabled = parsedConfig[:"controlplane-cluster-autoscaler"]
      puts "config::Using configmap scrape settings for controlplane-cluster-autoscaler: #{@controlplane_cluster_autoscaler_enabled}"
    end
    if !parsedConfig[:"controlplane-etcd"].nil?
      @controlplane_etcd_enabled = parsedConfig[:"controlplane-etcd"]
      puts "config::Using configmap scrape settings for controlplane-etcd: #{@controlplane_etcd_enabled}"
    end
    if !parsedConfig[:"controlplane-prometheuscollectorhealth"].nil?
      @controleplane_prometheuscollectorhealth_enabled = parsedConfig[:"controlplane-prometheuscollectorhealth"]
      puts "config::Using configmap scrape settings for controlplane_prometheuscollectorhealth: #{@controleplane_prometheuscollectorhealth_enabled}"
    end

    if ENV["MODE"].nil? && ENV["MODE"].strip.downcase == "advanced"
      controllerType = ENV["CONTROLLER_TYPE"]
      if controllerType == "ReplicaSet" && ENV["OS_TYPE"].downcase == "linux" && !@controlplane_kube_controller_manager_enabled && !@controlplane_kube_scheduler_enabled && !@controlplane_apiserver_enabled && !@controlplane_cluster_autoscaler_enabled && !@controlplane_etcd_enabled
        @noDefaultsEnabled = true
      end
    elsif !@controlplane_kube_controller_manager_enabled && !@controlplane_kube_scheduler_enabled && !@controlplane_apiserver_enabled && !@controlplane_cluster_autoscaler_enabled && !@controlplane_etcd_enabled
      @noDefaultsEnabled = true
    end
    if @noDefaultsEnabled
      ConfigParseErrorLogger.logWarning(LOGGING_PREFIX, "No default scrape configs enabled")
    end
  rescue => errorStr
    ConfigParseErrorLogger.logError(LOGGING_PREFIX, "Exception while reading config map settings for default scrape settings - #{errorStr}, using defaults, please check config map for errors")
  end
end

@configSchemaVersion = ENV["AZMON_AGENT_CFG_SCHEMA_VERSION"]
ConfigParseErrorLogger.logSection(LOGGING_PREFIX, "Start default-scrape-settings Processing")
# set default targets for MAC mode
if !ENV['MAC'].nil? && !ENV['MAC'].empty? && ENV['MAC'].strip.downcase == "true"
  ConfigParseErrorLogger.logWarning(LOGGING_PREFIX, "MAC mode is enabled. Only enabling targets controlplane_apiserver_enabled, controlplane_etcd_enabled for linux before config map processing....")
  @controlplane_kube_controller_manager_enabled = false
  @controlplane_kube_scheduler_enabled = false
  @controlplane_cluster_autoscaler_enabled = false
  @controleplane_prometheuscollectorhealth_enabled = false
end
if !@configSchemaVersion.nil? && !@configSchemaVersion.empty? && @configSchemaVersion.strip.casecmp("v1") == 0 #note v1 is the only supported schema version, so hardcoding it
  configMapSettings = parseConfigMap
  if !configMapSettings.nil?
    populateSettingValuesFromConfigMap(configMapSettings)
  end
else
  if (File.file?(@configMapMountPath))
    ConfigParseErrorLogger.logError(LOGGING_PREFIX, "Unsupported/missing config schema version - '#{@configSchemaVersion}' , using defaults, please use supported schema version")
  end
end

# Write the settings to file, so that they can be set as environment variables
file = File.open("/opt/microsoft/configmapparser/config_default_scrape_settings_env_var", "w")

if !file.nil?
  file.write("AZMON_PROMETHEUS_CONTROLPLANE_KUBE_CONTROLLER_MANAGER_ENABLED=#{@controlplane_kube_controller_manager_enabled}\n")
  file.write("AZMON_PROMETHEUS_CONTROLPLANE_KUBE_SCHEDULER_ENABLED=#{@controlplane_kube_scheduler_enabled}\n")
  file.write("AZMON_PROMETHEUS_CONTROLPLANE_APISERVER_ENABLED=#{@controlplane_apiserver_enabled}\n")
  file.write("AZMON_PROMETHEUS_CONTROLPLANE_CLUSTER_AUTOSCALER_ENABLED=#{@controlplane_cluster_autoscaler_enabled}\n")
  file.write("AZMON_PROMETHEUS_CONTROLPLANE_ETCD_ENABLED=#{@controlplane_etcd_enabled}\n")
  file.write("AZMON_PROMETHEUS_CONTROLPLANE_COLLECTOR_HEALTH_SCRAPING_ENABLED=#{@controleplane_prometheuscollectorhealth_enabled}\n")
  file.write("AZMON_PROMETHEUS_NO_DEFAULT_SCRAPING_ENABLED=#{@noDefaultsEnabled}\n")
  # Close file after writing all metric collection setting environment variables
  file.close
else
  ConfigParseErrorLogger.logError(LOGGING_PREFIX, "Exception while opening file for writing default-scrape-settings config environment variables")
end
ConfigParseErrorLogger.logSection(LOGGING_PREFIX, "End default-scrape-settings Processing")

File: otelcollector/shared/tomlparser-ccp-default-scrape-settings.go
package main

import (
	"fmt"
	"os"
	"regexp"
	"strings"
)

func (fcl *FilesystemConfigLoader) SetDefaultScrapeSettings() (map[string]string, error) {
	config := make(map[string]string)
	// Set default values
	config["controlplane-apiserver"] = "true"
	config["controlplane-cluster-autoscaler"] = "false"
	config["controlplane-kube-scheduler"] = "false"
	config["controlplane-kube-controller-manager"] = "false"
	config["controlplane-etcd"] = "true"
	return config, nil
}

func (fcl *FilesystemConfigLoader) ParseConfigMapForDefaultScrapeSettings() (map[string]string, error) {
	config := make(map[string]string)
	// Set default values
	config["controlplane-apiserver"] = "true"
	config["controlplane-cluster-autoscaler"] = "false"
	config["controlplane-kube-scheduler"] = "false"
	config["controlplane-kube-controller-manager"] = "false"
	config["controlplane-etcd"] = "true"

	if _, err := os.Stat(fcl.ConfigMapMountPath); os.IsNotExist(err) {
		fmt.Println("configmap for ccp default scrape settings not mounted, using defaults")
		return config, nil
	}

	content, err := os.ReadFile(string(fcl.ConfigMapMountPath))
	if err != nil {
		return config, fmt.Errorf("using default values, error reading config map file: %s", err)
	}

	lines := strings.Split(string(content), "\n")
	for _, line := range lines {
		parts := strings.SplitN(line, "=", 2)
		if len(parts) == 2 {
			config[strings.TrimSpace(parts[0])] = strings.TrimSpace(parts[1])
		}
	}

	fmt.Println("using configmap for ccp scrape settings...")
	return config, nil
}

func (cp *ConfigProcessor) PopulateSettingValues(parsedConfig map[string]string) {
	if val, ok := parsedConfig["controlplane-kube-controller-manager"]; ok && val != "" {
		cp.ControlplaneKubeControllerManager = val
		fmt.Printf("config::Using scrape settings for controlplane-kube-controller-manager: %v\n", cp.ControlplaneKubeControllerManager)
	}

	if val, ok := parsedConfig["controlplane-kube-scheduler"]; ok && val != "" {
		cp.ControlplaneKubeScheduler = val
		fmt.Printf("config::Using scrape settings for controlplane-kube-scheduler: %v\n", cp.ControlplaneKubeScheduler)
	}

	if val, ok := parsedConfig["controlplane-apiserver"]; ok && val != "" {
		cp.ControlplaneApiserver = val
		fmt.Printf("config::Using scrape settings for controlplane-apiserver: %v\n", cp.ControlplaneApiserver)
	}

	if val, ok := parsedConfig["controlplane-cluster-autoscaler"]; ok && val != "" {
		cp.ControlplaneClusterAutoscaler = val
		fmt.Printf("config::Using scrape settings for controlplane-cluster-autoscaler: %v\n", cp.ControlplaneClusterAutoscaler)
	}

	if val, ok := parsedConfig["controlplane-etcd"]; ok && val != "" {
		cp.ControlplaneEtcd = val
		fmt.Printf("config::Using scrape settings for controlplane-etcd: %v\n", cp.ControlplaneEtcd)
	}

	if os.Getenv("MODE") == "" && strings.ToLower(strings.TrimSpace(os.Getenv("MODE"))) == "advanced" {
		controllerType := os.Getenv("CONTROLLER_TYPE")
		if controllerType == "ReplicaSet" && strings.ToLower(os.Getenv("OS_TYPE")) == "linux" &&
			cp.ControlplaneKubeControllerManager == "" && cp.ControlplaneKubeScheduler == "" &&
			cp.ControlplaneApiserver == "" && cp.ControlplaneClusterAutoscaler == "" && cp.ControlplaneEtcd == "" {
			cp.NoDefaultsEnabled = true
		}
	} else if cp.ControlplaneKubeControllerManager == "" && cp.ControlplaneKubeScheduler == "" &&
		cp.ControlplaneApiserver == "" && cp.ControlplaneClusterAutoscaler == "" && cp.ControlplaneEtcd == "" {
		cp.NoDefaultsEnabled = true
	}

	if cp.NoDefaultsEnabled {
		fmt.Printf("No default scrape configs enabled")
	}
}

func (fcw *FileConfigWriter) WriteDefaultScrapeSettingsToFile(filename string, cp *ConfigProcessor) error {
	file, err := os.Create(filename)
	if err != nil {
		return fmt.Errorf("Exception while opening file for writing prometheus-collector config environment variables: %s", err)
	}
	defer file.Close()

	file.WriteString(fmt.Sprintf("AZMON_PROMETHEUS_CONTROLPLANE_KUBE_CONTROLLER_MANAGER_ENABLED=%v\n", cp.ControlplaneKubeControllerManager))
	file.WriteString(fmt.Sprintf("AZMON_PROMETHEUS_CONTROLPLANE_KUBE_SCHEDULER_ENABLED=%v\n", cp.ControlplaneKubeScheduler))
	file.WriteString(fmt.Sprintf("AZMON_PROMETHEUS_CONTROLPLANE_APISERVER_ENABLED=%v\n", cp.ControlplaneApiserver))
	file.WriteString(fmt.Sprintf("AZMON_PROMETHEUS_CONTROLPLANE_CLUSTER_AUTOSCALER_ENABLED=%v\n", cp.ControlplaneClusterAutoscaler))
	file.WriteString(fmt.Sprintf("AZMON_PROMETHEUS_CONTROLPLANE_ETCD_ENABLED=%v\n", cp.ControlplaneEtcd))
	file.WriteString(fmt.Sprintf("AZMON_PROMETHEUS_NO_DEFAULT_SCRAPING_ENABLED=%v\n", cp.NoDefaultsEnabled))

	return nil
}

func (c *Configurator) ConfigureDefaultScrapeSettings() {
	configSchemaVersion := os.Getenv("AZMON_AGENT_CFG_SCHEMA_VERSION")

	fmt.Printf("Start prometheus-collector-settings Processing\n")

	if configSchemaVersion != "" && strings.TrimSpace(configSchemaVersion) == "v1" {
		configMapSettings, err := c.ConfigLoader.ParseConfigMapForDefaultScrapeSettings()
		if err == nil && len(configMapSettings) > 0 {
			c.ConfigParser.PopulateSettingValues(configMapSettings)
		}
	} else {
		defaultSettings, err := c.ConfigLoader.SetDefaultScrapeSettings()
		if err == nil && len(defaultSettings) > 0 {
			c.ConfigParser.PopulateSettingValues(defaultSettings)
		}
		fmt.Printf("Unsupported/missing config schema version - '%s', using defaults, please use supported schema version\n", configSchemaVersion)
	}

	if mac := os.Getenv("MAC"); mac != "" && strings.TrimSpace(mac) == "true" {
		clusterArray := strings.Split(strings.TrimSpace(os.Getenv("CLUSTER")), "/")
		c.ConfigParser.ClusterAlias = clusterArray[len(clusterArray)-1]
	} else {
		c.ConfigParser.ClusterAlias = os.Getenv("CLUSTER")
	}

	if c.ConfigParser.ClusterAlias != "" && len(c.ConfigParser.ClusterAlias) > 0 {
		// replace all non alpha-numeric characters with "_"  -- this is to ensure that all down stream places where this is used (like collector, telegraf config etc are keeping up with sanity)
		c.ConfigParser.ClusterAlias = regexp.MustCompile(`[^0-9a-zA-Z]+`).ReplaceAllString(c.ConfigParser.ClusterAlias, "_")
		c.ConfigParser.ClusterAlias = strings.Trim(c.ConfigParser.ClusterAlias, "_")
		fmt.Printf("After replacing non-alpha-numeric characters with '_': %s\n", c.ConfigParser.ClusterAlias)
	}

	err := c.ConfigWriter.WriteDefaultScrapeSettingsToFile(c.ConfigFilePath, c.ConfigParser)
	if err != nil {
		fmt.Printf("%v\n", err)
		return
	}

	fmt.Printf("End prometheus-collector-settings Processing\n")
}

func tomlparserCCPDefaultScrapeSettings() {
	configurator := &Configurator{
		ConfigLoader:   &FilesystemConfigLoader{ConfigMapMountPath: "/etc/config/settings/default-scrape-settings-enabled"},
		ConfigWriter:   &FileConfigWriter{Config: map[string]string{}},
		ConfigFilePath: "/opt/microsoft/configmapparser/config_default_scrape_settings_env_var",
		ConfigParser:   &ConfigProcessor{},
	}

	fmt.Println("Start ccp-default-scrape-settings Processing")
	configurator.ConfigureDefaultScrapeSettings()
	fmt.Println("End ccp-default-scrape-settings Processing")
}
